### ISSUE TITLE
Potential Issue in `NetworkState::announce_new_block()` Leading to Non-Random Peer Selection

### ISSUE THREAD

**Issue by Kuly14** [paradigmxyz/reth#5906]
### Describe the bug

In the [NetworkState::announce_new_block()](https://github.com/paradigmxyz/reth/blob/main/crates/net/network/src/state.rs#L172) function, we currently select a subset of peers for block announcement by taking the square root of the total number of peers and then iterating over the `active_peers` HashMap to send the new block to the first `sqrt(number of peers)` peers.

The issue arises from the stable ordering of peers in the `HashMap`. If there are no significant changes in the peer list over extended periods, the function consistently selects the same set of peers for block propagation. 

Even with occasional changes in the peer list, if the total number of peers is relatively stable (e.g., around 25), the beginning portion of the HashMap (from where peers are selected) may remain largely unchanged. Consequently, this could lead to prolonged periods where the same nodes are repeatedly chosen for gossiping new blocks.

This could **theoretically** contribute to the formation of network partitions, albeit likely temporary.

I may be missing something important here, so I'm eager to hear if anyone has additional insights or perspectives on this matter.

### Steps to reproduce

When you run this code multiple times, the order of the first ~5 elements in the HashMap often seems to stay the same, though this consistency isn't guaranteed by the data structure and can vary. But more often than not the order stays pretty much the same.

```rust
use std::collections::HashMap;

fn main() {
    let mut map = HashMap::new();
    for i in 1..=25 {
        map.insert(i, 1);
    }
    for (i, _) in &map {
        print!("{}", i);
    }
    println!();

    map.shrink_to_fit();
    
    map.insert(26, 1);

    for (i, _) in &map {
        print!("{}", i);
    }
    println!()
}
```

### What type of node are you running?

None

### Code of Conduct

- [X] I agree to follow the Code of Conduct

**Comment by Rjected** [comment]
This is also observed in the txpool: https://github.com/paradigmxyz/reth/pull/5310

Block announcements are disabled post-merge so this is unlikely to impact real world usage. the fix for txpool is likely to be applicable to this code as well

**Comment by mattsse** [comment]
yeah this would be an issue if block propagation would still be over devp2p, so this has low prio but is a valid issue and should be simple to fix

**Comment by cetceeve** [comment]
Hi, I have wanted to check out some more low-level blockchain projects recently. This seems like a good issue to get started and look into the code a bit. If it's fine for you I would take this. 😃

**Comment by Kuly14** [comment]
@mattsse I thought the optimism feature still used this but looking at it now it seems like they also handle block propagation in the CL. Still may be relevant for BSC and polygon tho.

**Comment by mattsse** [comment]
for BSC yes, one of reasons why i have reserved this functionality at all
